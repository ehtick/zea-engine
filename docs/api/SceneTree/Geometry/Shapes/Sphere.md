<a name="Sphere"></a>

## Sphere ⇐ <code>Mesh</code>
A class for generating a sphere geometry.

**Kind**: global class  
**Extends**: <code>Mesh</code>  
<a name="new_Sphere_new"></a>

### new Sphere
Create a sphere.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| radius | <code>number</code> | <code>1</code> | The radius of the sphere. |
| sides | <code>number</code> | <code>12</code> | The number of sides. |
| loops | <code>number</code> | <code>12</code> | The number of loops. |

