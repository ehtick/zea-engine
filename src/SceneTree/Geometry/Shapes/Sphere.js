import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Mesh } from '../Mesh.js';


class Sphere extends Mesh {
    constructor(radius = 1.0, sides = 12, loops = 12) {
        super();

        if(isNaN(radius) || isNaN(sides) || isNaN(loops))
            throw("Invalid geom args");

        this.__radius = radius;
        this.__sides = (sides >= 3) ? sides : 3;
        this.__loops = (loops >= 3) ? loops : 3;

        this.addVertexAttribute('texCoords', Vec2);
        this.addVertexAttribute('normals', Vec3);
        this.__rebuild();
    }

    get radius() {
        return this.__radius
    }

    set radius(val) {
        this.__radius = val;
        this.__resize();
    }

    get sides() {
        return this.__sides
    }

    set sides(val) {
        this.__sides = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    get loops() {
        return this.__loops
    }

    set loops(val) {
        this.__loops = (val >= 3) ? val : 3;
        this.__rebuild();
    }

    __rebuild() {
        let numVertices = 2 + this.__sides * this.__loops;
        let numTris = this.__sides * 2;
        let numQuads = this.__sides * this.__loops;
        this.setNumVertices(numVertices);
        this.setFaceCounts([numTris, numQuads]);

        //////////////////////////////
        // Set Vertex Positions

        let normals = this.getVertexAttribute('normals');
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, 1.0);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                normals.getValueRef(vertex).setFromOther(normal);
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        normals.getValueRef(vertex).set(0.0, 0.0, -1.0);
        vertex++;

        //////////////////////////////
        // build the topology
        let texCoords = this.getVertexAttribute('texCoords');

        // build the fan at the first pole.
        let faceIndex = 0;
        for (let j = 0; j < this.__sides; j++) {
            let v0 = 0;
            let v1 = ((j + 1) % this.__sides) + 1;
            let v2 = j + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(0.5, 0.0);
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), 0.0);
            let uv2 = new Vec2(1.0 - (j / this.__sides), 1.0 / (this.__loops + 1));
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }
        // build the fan at the second pole.
        for (let j = 0; j < this.__sides; j++) {
            let v0 = numVertices - 1;
            let v1 = (this.__sides * (this.__loops - 1)) + j + 1;
            let v2 = (this.__sides * (this.__loops - 1)) + ((j + 1) % this.__sides) + 1;
            this.setFaceVertexIndices(faceIndex, v0, v1, v2);

            let uv0 = new Vec2(1.0 - (j / this.__sides), this.__loops / (this.__loops + 1));
            let uv1 = new Vec2(1.0 - ((j + 1) / this.__sides), this.__loops / (this.__loops + 1));
            let uv2 = new Vec2(0.5, 1.0);
            texCoords.setFaceVertexValue(faceIndex, 0, uv0);
            texCoords.setFaceVertexValue(faceIndex, 1, uv1);
            texCoords.setFaceVertexValue(faceIndex, 2, uv2);

            faceIndex++;
        }

        for (let i = 0; i < this.__loops - 1; i++) {
            for (let j = 0; j < this.__sides; j++) {
                let v0 = (this.__sides * i) + j + 1;
                let v1 = (this.__sides * i) + ((j + 1) % this.__sides) + 1;
                let v2 = (this.__sides * (i + 1)) + ((j + 1) % this.__sides) + 1;
                let v3 = (this.__sides * (i + 1)) + j + 1;
                this.setFaceVertexIndices(faceIndex, v0, v1, v2, v3);

                texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / this.__loops, j / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / this.__loops, (j + 1) / this.__loops));
                texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / this.__loops, j / this.__loops));
                faceIndex++;
            }
        }

        this.setBoundingBoxDirty();
    }

    __resize() {
        //////////////////////////////
        // Set Vertex Positions
        let vertex = 0;
        let normal = new Vec3(0.0, 0.0, 1.0);
        this.getVertex(vertex).set(0.0, 0.0, this.__radius);
        vertex++;

        for (let i = 0; i < this.__loops; i++) {
            let theta = ((i + 1) / (this.__loops + 1)) * Math.PI;
            for (let j = 0; j < this.__sides; j++) {
                let phi = (j / this.__sides) * 2.0 * Math.PI;
                normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta));

                // Set positions and normals at the same time.
                this.getVertex(vertex).setFromOther(normal.scale(this.__radius));
                vertex++;
            }
        }
        this.getVertex(vertex).set(0.0, 0.0, -this.__radius);
        vertex++;

        this.setBoundingBoxDirty();
    }

    toJSON() {
        let json = super.toJSON();
        json['x'] = this.__x;
        json['y'] = this.__y;
        json['z'] = this.__z;
        return json
    }
};

export {
    Sphere
};
//export default Sphere;