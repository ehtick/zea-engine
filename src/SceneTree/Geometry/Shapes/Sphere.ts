import { Vec2, Vec3 } from '../../../Math/index'
import { NumberParameter } from '../../Parameters/NumberParameter'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * A class for generating a sphere geometry.
 *
 * ```
 * const sphere = new Sphere(1.4, 13)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Radius of the sphere.
 * * **Sides(`NumberParameter`):** Specifies the number of subdivisions around the `Z` axis.
 * * **Loops(`NumberParameter`):** Specifies the number of subdivisions(stacks) along the `Z` axis.
 *
 * @extends {ProceduralMesh}
 */
class Sphere extends ProceduralMesh {
  /**
   * @member {NumberParameter} loopsParam - Specifies the number of subdivisions(stacks) along the `Z` axis.
   */
  loopsParam: NumberParameter

  /**
   * @member {NumberParameter} radiusParam - Radius of the sphere.
   */
  radiusParam: NumberParameter

  /**
   * @member {NumberParameter} sidesParam - Specifies the number of subdivisions around the `Z` axis.
   */
  sidesParam: NumberParameter

  /**
   * Creates an instance of Sphere.
   * @param {number} [radius=1.0] - The radius of the sphere.
   * @param {number} [sides=12] - The number of sides.
   * @param {number} [loops=12] - The number of loops.
   * @param {boolean} [addNormals=true] - Compute vertex normals for the geometry
   * @param {boolean} [addTextureCoords=true] - Compute texture coordinates for the geometry
   */
  constructor(radius = 1.0, sides = 12, loops = 12) {
    super()

    if (isNaN(radius) || isNaN(sides) || isNaN(loops)) throw new Error('Invalid geom args')

    this.radiusParam = this.addParameter(new NumberParameter('Radius', radius)) as NumberParameter
    this.sidesParam = this.addParameter(
      new NumberParameter('Sides', sides >= 3 ? sides : 3, [3, 200], 1)
    ) as NumberParameter
    this.loopsParam = this.addParameter(
      new NumberParameter('Loops', loops >= 3 ? loops : 3, [3, 200], 1)
    ) as NumberParameter

    this.addVertexAttribute('texCoords', new Vec2Attribute())
    this.addVertexAttribute('normals', new Vec3Attribute())

    this.topologyParams.push('Sides')
    this.topologyParams.push('Loops')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const radius = this.radiusParam.value || 1.0
    const nbSides = this.sidesParam.value || 12
    const nbLoops = this.loopsParam.value || 12

    const numVertices = 2 + nbSides * nbLoops
    const numTris = nbSides * 2
    const numQuads = nbSides * nbLoops
    this.setNumVertices(numVertices)
    this.setFaceCounts([numTris, numQuads])

    // ////////////////////////////
    // Set Vertex Positions

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    const normal = new Vec3(0.0, 0.0, 1.0)
    let vertex = 0
    if (!positions || !normals) return

    positions.getValueRef(vertex).set(0.0, 0.0, radius)
    normals.getValueRef(vertex).set(0.0, 0.0, 1.0)
    vertex++

    for (let i = 0; i < nbLoops; i++) {
      const theta = ((i + 1) / (nbLoops + 1)) * Math.PI
      for (let j = 0; j < nbSides; j++) {
        const phi = -((j / nbSides) * 2.0 * Math.PI)
        normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta))

        // Set positions and normals at the same time.
        positions.getValueRef(vertex).setFromOther(normal.scale(radius))
        normals.getValueRef(vertex).setFromOther(normal)
        vertex++
      }
    }

    positions.getValueRef(vertex).set(0.0, 0.0, -radius)
    normals.getValueRef(vertex).set(0.0, 0.0, -1.0)
    vertex++

    // ////////////////////////////
    // Build the topology
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      // build the fan at the first pole.
      let faceIndex = 0
      for (let j = 0; j < nbSides; j++) {
        const v0 = 0
        const v1 = ((j + 1) % nbSides) + 1
        const v2 = j + 1
        this.setFaceVertexIndices(faceIndex, [v0, v1, v2])

        if (texCoords) {
          const uv0 = new Vec2(0.5, 0.0)
          const uv1 = new Vec2(1.0 - (j + 1) / nbSides, 0.0)
          const uv2 = new Vec2(1.0 - j / nbSides, 1.0 / (nbLoops + 1))
          texCoords.setFaceVertexValue(faceIndex, 0, uv0)
          texCoords.setFaceVertexValue(faceIndex, 1, uv1)
          texCoords.setFaceVertexValue(faceIndex, 2, uv2)
        }

        faceIndex++
      }

      // Build the fan at the second pole.
      for (let j = 0; j < nbSides; j++) {
        const v0 = numVertices - 1
        const v1 = nbSides * (nbLoops - 1) + j + 1
        const v2 = nbSides * (nbLoops - 1) + ((j + 1) % nbSides) + 1
        this.setFaceVertexIndices(faceIndex, [v0, v1, v2])

        const uv0 = new Vec2(1.0 - j / nbSides, nbLoops / (nbLoops + 1))
        const uv1 = new Vec2(1.0 - (j + 1) / nbSides, nbLoops / (nbLoops + 1))
        const uv2 = new Vec2(0.5, 1.0)
        texCoords.setFaceVertexValue(faceIndex, 0, uv0)
        texCoords.setFaceVertexValue(faceIndex, 1, uv1)
        texCoords.setFaceVertexValue(faceIndex, 2, uv2)

        faceIndex++
      }

      for (let i = 0; i < nbLoops - 1; i++) {
        for (let j = 0; j < nbSides; j++) {
          const v0 = nbSides * i + j + 1
          const v1 = nbSides * i + ((j + 1) % nbSides) + 1
          const v2 = nbSides * (i + 1) + ((j + 1) % nbSides) + 1
          const v3 = nbSides * (i + 1) + j + 1
          this.setFaceVertexIndices(faceIndex, [v0, v1, v2, v3])

          texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / nbLoops, j / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / nbLoops, j / nbLoops))
          faceIndex++
        }
      }
    }
  }

  /**
   * The resize method.
   * @private
   */
  resize() {
    const radius = this.radiusParam.value
    const nbSides = this.sidesParam.value
    const nbLoops = this.loopsParam.value
    if (!radius || !nbSides || !nbLoops) {
      console.warn('resize() failed')
      return
    }
    // ////////////////////////////
    // Set Vertex Positions
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    let vertex = 0
    const normal = new Vec3(0.0, 0.0, 1.0)
    positions.getValueRef(vertex).set(0.0, 0.0, radius)
    if (normals) normals.getValueRef(vertex).set(0.0, 0.0, 1.0)
    vertex++

    for (let i = 0; i < nbLoops; i++) {
      const theta = ((i + 1) / (nbLoops + 1)) * Math.PI
      for (let j = 0; j < nbSides; j++) {
        const phi = -((j / nbSides) * 2.0 * Math.PI)
        normal.set(Math.sin(theta) * Math.cos(phi), Math.sin(theta) * Math.sin(phi), Math.cos(theta))

        // Set positions and normals at the same time.
        positions.getValueRef(vertex).setFromOther(normal.scale(radius))
        if (normals) normals.getValueRef(vertex).setFromOther(normal)
        vertex++
      }
    }
    positions.getValueRef(vertex).set(0.0, 0.0, -radius)
    if (normals) normals.getValueRef(vertex).set(0.0, 0.0, -1.0)
    vertex++
  }
}

Registry.register('Sphere', Sphere)

export { Sphere }
