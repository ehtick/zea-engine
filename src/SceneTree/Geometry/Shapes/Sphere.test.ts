import { Sphere } from './Sphere'

describe('Sphere', () => {
  test('test default parameters.', () => {
    const sphere = new Sphere()

    expect(sphere.getParameter('Radius').getValue()).toBe(1.0)

    expect(sphere.getParameter('Sides').getValue()).toBe(12)

    expect(sphere.getParameter('Loops').getValue()).toBe(12)
  })

  test('test update parameters.', () => {
    const sphere = new Sphere()
    sphere.getParameter('Radius').setValue(2.6)
    sphere.getParameter('Sides').setValue(24)
    sphere.getParameter('Loops').setValue(30)

    expect(sphere.getParameter('Radius').getValue()).toBe(2.6)

    expect(sphere.getParameter('Sides').getValue()).toBe(24)

    expect(sphere.getParameter('Loops').getValue()).toBe(30)
  })

  it('saves to JSON (serialization).', () => {
    const sphere = new Sphere(2.6, 5, 5)

    const outputJSON = sphere.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const sphere = new Sphere()
    const inputJSON = {
      params: {
        Loops: {
          range: [3, 200],
          step: 1,
          value: 5,
        },
        Radius: {
          value: 2.6,
        },
        Sides: {
          range: [3, 200],
          step: 1,
          value: 5,
        },
      },
      type: 'Sphere',
      vertexAttributes: {},
    }

    sphere.fromJSON(inputJSON)

    expect(sphere.toJSON()).toEqual(new Sphere(2.6, 5, 5).toJSON())
  })
})
