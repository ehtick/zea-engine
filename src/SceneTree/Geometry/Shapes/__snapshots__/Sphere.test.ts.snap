// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sphere saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "Loops": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 5,
    },
    "Radius": Object {
      "value": 2.6,
    },
    "Sides": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 5,
    },
  },
  "type": "Sphere",
  "vertexAttributes": Object {},
}
`;
